mllib/src/main/scala/org/apache/spark/ml/tree/impl/RandomForest.scala

```diff
@@ -590,8 +599,8 @@ private[spark] object RandomForest extends Logging {
         if (!isLeaf) {
           node.split = Some(split)
           val childIsLeaf = (LearningNode.indexToLevel(nodeIndex) + 1) == metadata.maxDepth
-          val leftChildIsLeaf = childIsLeaf || (stats.leftImpurity == 0.0)
-          val rightChildIsLeaf = childIsLeaf || (stats.rightImpurity == 0.0)
+          val leftChildIsLeaf = childIsLeaf || (math.abs(stats.leftImpurity) < 1e-16)
+          val rightChildIsLeaf = childIsLeaf || (math.abs(stats.rightImpurity) < 1e-16)
```

Comment:
the code for left/right child looks very similar, consider refactoring to a function.
also, should 1e-16 be moved to a constant, or is there a global constant somewhere for this (or how was this value chosen)?